src/main/java/pl/psnc/dei/service/BatchService.java

```diff
@@ -314,4 +306,88 @@ private String extractIdentifier(String key) {
 		Hibernate.initialize(currentProject.getDatasets());
 		return currentProject.getDatasets();
 	}
-}
+
+	/**
+	 * Saves records to DB, and calculate difference of saved and provided ones, as some of them can
+	 * be omitted during persiting process
+	 * @param projectName name of project to which data should be saved
+	 * @param datasetName name of dataset from which data comes
+	 * @param recordsIds id of data to save
+	 * @return saved records
+	 * @throws NotFoundException
+	 */
+	public Set<Record> uploadRecordsToProject(String projectName,
+											  String datasetName,
+											  Set<String> recordsIds)
+			throws NotFoundException {
+		String datasetId = null;
+		if (datasetName != null) {
+			Optional<Dataset> optionalDataset = this.datasetsReposotory.findByName(datasetName);
+			if (optionalDataset.isEmpty()) {
+				throw new NotFoundException("Dataset with name: " + datasetName + " not found");
+			}
+			datasetId = optionalDataset.get().getDatasetId();
+		}
+		Set<Record> records = this.uploadRecords(projectName, datasetId, recordsIds);
+		if (records.size() < recordsIds.size()) {
+			String difference = records
+					.stream()
+					.map(Record::getIdentifier)
+					.filter(identifier -> !recordsIds.contains(identifier)).collect(Collectors.joining(","));
+			if (difference.isEmpty()) {
+				difference = String.join(",", recordsIds);
+			}
+			log.warn("Following records will not be added to the import: {}", difference);
+		}
+		return records;
+	}
+
+	public List<Import> makeComplexImport(MultipartFile file, String name, String projectName, String datasetName) throws IOException, NotFoundException {
+		File saved = null;
+		try {
+			saved = File.createTempFile("tmp", ".csv");
+			file.transferTo(saved);
+			return this.makeComplexImport(saved, name, projectName, datasetName);
+		} catch (Exception e) {
```

Comment:
If the only one purpose of `catch` block is to rethrow exception, could you not just omit it?

Reply:
I could, `catch` has no special purpose here, but `try/catch/finally` is more readable for me